server/container/guice/guice-common/src/main/java/org/apache/james/JamesServerMain.java

```diff
@@ -19,20 +19,36 @@
 
 package org.apache.james;
 
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.apache.commons.configuration2.ex.ConfigurationException;
 import org.apache.james.server.core.configuration.Configuration;
+import org.apache.james.utils.PropertiesProvider;
 
+import com.google.common.collect.ImmutableList;
 import com.google.inject.Module;
+import com.google.inject.util.Modules;
 
 public interface JamesServerMain {
+    @FunctionalInterface
+    interface ConfiguredModulesSupplier {
+        Stream<Module> configuredModules(PropertiesProvider propertiesProvider) throws ConfigurationException;
+    }
+
     static void main(Module... modules) throws Exception {
         Configuration configuration = Configuration.builder()
             .useWorkingDirectoryEnvProperty()
             .build();
 
+        main(configuration, ImmutableList.copyOf(modules));
+    }
+
+    static void main(Configuration configuration, List<Module> baseModules) throws Exception {
```

Comment:
why `base`?
and why do you want to expose this main method?

Reply:
Lets call it modules.

The goal here is to avoid products to instanciate congiguration metchod two tomes.

The above method (with only module) should imo eventually be removed.